File: egdc/mog-solver/src/lexer.rs
//! Mog lexer -- tokenizes source code into a token stream.

use crate::token::{keyword_lookup, Token, TokenKind};

/// Lexing error.
#[derive(Debug)]
pub struct LexError {
    pub msg: String,
    pub line: u32,
    pub col: u32,
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "Lex error at L{}:{}: {}", self.line, self.col, self.msg)
    }
}

impl std::error::Error for LexError {}

/// Tokenizes Mog source code.
pub struct Lexer {
    src: Vec<char>,
    pos: usize,
    line: u32,
    col: u32,
    tokens: Vec<Token>,
}

impl Lexer {
    pub fn new(source: &str) -> Self {
        Lexer {
            src: source.chars().collect(),
            pos: 0,
            line: 1,
            col: 1,
            tokens: Vec::new(),
        }
    }

    fn ch(&self) -> char {
        if self.pos < self.src.len() {
            self.src[self.pos]
        } else {
            '\0'
        }
    }

    fn peek_offset(&self, offset: usize) -> char {
        let p = self.pos + offset;
        if p < self.src.len() {
            self.src[p]
        } else {
            '\0'
        }
    }

    fn advance(&mut self) -> char {
        let ch = self.ch();
        self.pos += 1;
        if ch == '\n' {
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        ch
    }

    fn emit(&mut self, kind: TokenKind, value: impl Into<String>, line: u32, col: u32) {
        self.tokens.push(Token::new(kind, value, line, col));
    }

    fn skip_whitespace_and_comments(&mut self) {
        while self.pos < self.src.len() {
            let ch = self.ch();
            if ch == ' ' || ch == '\t' || ch == '\r' || ch == '\n' {
                self.advance();
            } else if ch == '/' && self.peek_offset(1) == '/' {
                // Line comment
                while self.pos < self.src.len() && self.ch() != '\n' {
                    self.advance();
                }
            } else if ch == '/' && self.peek_offset(1) == '*' {
                // Block comment
                self.advance(); // /
                self.advance(); // *
                let mut depth: i32 = 1;
                while self.pos < self.src.len() && depth > 0 {
                    if self.ch() == '*' && self.peek_offset(1) == '/' {
                        self.advance();
                        self.advance();
                        depth -= 1;
                    } else if self.ch() == '/' && self.peek_offset(1) == '*' {
                        self.advance();
                        self.advance();
                        depth += 1;
                    } else {
                        self.advance();
                    }
                }
            } else {
                break;
            }
        }
    }

    fn read_string(&mut self, _is_fstring: bool) -> String {
        let mut parts = Vec::new();
        while self.pos < self.src.len() && self.ch() != '"' {
            if self.ch() == '\\' {
                self.advance();
                let esc = self.advance();
                match esc {
                    'n' => parts.push('\n'),
                    't' => parts.push('\t'),
                    '\\' => parts.push('\\'),
                    '"' => parts.push('"'),
                    '{' => parts.push('{'),
                    _ => {
                        parts.push('\\');
                        parts.push(esc);
                    }
                }
            } else {
                parts.push(self.advance());
            }
        }
        if self.pos < self.src.len() {
            self.advance(); // closing "
        }
        parts.into_iter().collect()
    }

    fn read_number(&mut self) -> (TokenKind, String) {
        let start = self.pos;
        let mut is_float = false;
        while self.pos < self.src.len() && (self.ch().is_ascii_digit() || self.ch() == '_') {
            self.advance();
        }
        if self.ch() == '.' && self.peek_offset(1).is_ascii_digit() {
            is_float = true;
            self.advance(); // .
            while self.pos < self.src.len() && (self.ch().is_ascii_digit() || self.ch() == '_') {
                self.advance();
            }
        }
        if self.ch() == 'e' || self.ch() == 'E' {
            is_float = true;
            self.advance();
            if self.ch() == '+' || self.ch() == '-' {
                self.advance();
            }
            while self.pos < self.src.len() && self.ch().is_ascii_digit() {
                self.advance();
            }
        }
        let val: String = self.src[start..self.pos].iter().filter(|c| **c != '_').collect();
        (if is_float { TokenKind::FloatLit } else { TokenKind::IntLit }, val)
    }

    fn read_ident(&mut self) -> String {
        let start = self.pos;
        while self.pos < self.src.len() && (self.ch().is_alphanumeric() || self.ch() == '_') {
            self.advance();
        }
        self.src[start..self.pos].iter().collect()
    }

    pub fn tokenize(&mut self) -> Vec<Token> {
        loop {
            self.skip_whitespace_and_comments();
            if self.pos >= self.src.len() {
                break;
            }

            let line = self.line;
            let col = self.col;
            let ch = self.ch();

            // F-strings
            if ch == 'f' && self.peek_offset(1) == '"' {
                self.advance(); // f
                self.advance(); // "
                let val = self.read_string(true);
                self.emit(TokenKind::FStringLit, val, line, col);
                continue;
            }

            // Identifiers and keywords
            if ch.is_alphabetic() || ch == '_' {
                let word = self.read_ident();
                let kind = keyword_lookup(&word).unwrap_or(TokenKind::Ident);
                self.emit(kind, word, line, col);
                continue;
            }

            // Numbers
            if ch.is_ascii_digit() {
                let (kind, val) = self.read_number();
                self.emit(kind, val, line, col);
                continue;
            }

            // Strings
            if ch == '"' {
                self.advance();
                let val = self.read_string(false);
                self.emit(TokenKind::StringLit, val, line, col);
                continue;
            }

            // Two-char operators
            let nch = self.peek_offset(1);
            let two = format!("{ch}{nch}");

            macro_rules! two_char_op {
                ($s:expr, $kind:expr) => {
                    if two == $s {
                        self.advance();
                        self.advance();
                        self.emit($kind, $s, line, col);
                        continue;
                    }
                };
            }

            two_char_op!(":=", TokenKind::ColonEq);
            two_char_op!("==", TokenKind::EqEq);
            two_char_op!("!=", TokenKind::BangEq);
            two_char_op!("<=", TokenKind::LtEq);
            two_char_op!(">=", TokenKind::GtEq);
            two_char_op!("->", TokenKind::Arrow);
            two_char_op!("=>", TokenKind::FatArrow);
            two_char_op!("&&", TokenKind::AmpAmp);
            two_char_op!("||", TokenKind::PipePipe);
            two_char_op!("<<", TokenKind::Shl);
            two_char_op!(">>", TokenKind::Shr);
            two_char_op!("**", TokenKind::StarStar);
            two_char_op!("..", TokenKind::DotDot);

            // Single-char
            self.advance();
            let kind = match ch {
                '+' => TokenKind::Plus,
                '-' => TokenKind::Minus,
                '*' => TokenKind::Star,
                '/' => TokenKind::Slash,
                '%' => TokenKind::Percent,
                '<' => TokenKind::Lt,
                '>' => TokenKind::Gt,
                '&' => TokenKind::Amp,
                '|' => TokenKind::Pipe,
                '^' => TokenKind::Caret,
                '!' => TokenKind::Bang,
                '?' => TokenKind::Question,
                '(' => TokenKind::LParen,
                ')' => TokenKind::RParen,
                '{' => TokenKind::LBrace,
                '}' => TokenKind::RBrace,
                '[' => TokenKind::LBracket,
                ']' => TokenKind::RBracket,
                ',' => TokenKind::Comma,
                ';' => TokenKind::Semicolon,
                ':' => TokenKind::Colon,
                '.' => TokenKind::Dot,
                '=' => TokenKind::Eq,
                '_' => TokenKind::Underscore,
                _ => {
                    return self.tokens.clone(); // Error will be apparent to parser
                }
            };
            self.emit(kind, ch.to_string(), line, col);
        }

        self.emit(TokenKind::Eof, "", self.line, self.col);
        self.tokens.clone()
    }
}

/// Tokenize Mog source code.
pub fn lex(source: &str) -> Result<Vec<Token>, LexError> {
    let mut lexer = Lexer::new(source);
    Ok(lexer.tokenize())
}

File: egdc/mog/lang/lexer.py
"""Mog lexer — tokenizes source code into a token stream."""
from __future__ import annotations
from .tokens import Token, TT, KEYWORDS


class LexError(Exception):
    def __init__(self, msg: str, line: int, col: int):
        super().__init__(f"Lex error at L{line}:{col}: {msg}")
        self.line, self.col = line, col


class Lexer:
    """Tokenizes Mog source code."""

    def __init__(self, source: str):
        self.src = source
        self.pos = 0
        self.line = 1
        self.col = 1
        self.tokens: list[Token] = []

    def _ch(self) -> str:
        return self.src[self.pos] if self.pos < len(self.src) else "\0"

    def _peek(self, offset: int = 1) -> str:
        p = self.pos + offset
        return self.src[p] if p < len(self.src) else "\0"

    def _advance(self) -> str:
        ch = self._ch()
        self.pos += 1
        if ch == "\n":
            self.line += 1
            self.col = 1
        else:
            self.col += 1
        return ch

    def _emit(self, tt: TT, value: str, line: int, col: int):
        self.tokens.append(Token(tt, value, line, col))

    def _skip_whitespace_and_comments(self):
        while self.pos < len(self.src):
            ch = self._ch()
            if ch in " \t\r\n":
                self._advance()
            elif ch == "/" and self._peek() == "/":
                # Line comment
                while self.pos < len(self.src) and self._ch() != "\n":
                    self._advance()
            elif ch == "/" and self._peek() == "*":
                # Block comment
                self._advance()  # /
                self._advance()  # *
                depth = 1
                while self.pos < len(self.src) and depth > 0:
                    if self._ch() == "*" and self._peek() == "/":
                        self._advance()
                        self._advance()
                        depth -= 1
                    elif self._ch() == "/" and self._peek() == "*":
                        self._advance()
                        self._advance()
                        depth += 1
                    else:
                        self._advance()
            else:
                break

    def _read_string(self, is_fstring: bool = False) -> str:
        """Read a string literal (opening " already consumed)."""
        parts = []
        while self.pos < len(self.src) and self._ch() != '"':
            if self._ch() == "\\":
                self._advance()
                esc = self._advance()
                if esc == "n":
                    parts.append("\n")
                elif esc == "t":
                    parts.append("\t")
                elif esc == "\\":
                    parts.append("\\")
                elif esc == '"':
                    parts.append('"')
                elif esc == "{":
                    parts.append("{")
                else:
                    parts.append("\\" + esc)
            else:
                parts.append(self._advance())
        if self.pos < len(self.src):
            self._advance()  # closing "
        return "".join(parts)

    def _read_number(self) -> tuple[TT, str]:
        """Read an integer or float literal."""
        start = self.pos
        is_float = False
        while self.pos < len(self.src) and (self._ch().isdigit() or self._ch() == "_"):
            self._advance()
        if self._ch() == "." and self._peek().isdigit():
            is_float = True
            self._advance()  # .
            while self.pos < len(self.src) and (self._ch().isdigit() or self._ch() == "_"):
                self._advance()
        if self._ch() in "eE":
            is_float = True
            self._advance()
            if self._ch() in "+-":
                self._advance()
            while self.pos < len(self.src) and self._ch().isdigit():
                self._advance()
        val = self.src[start:self.pos].replace("_", "")
        return (TT.FLOAT_LIT if is_float else TT.INT_LIT), val

    def _read_ident(self) -> str:
        start = self.pos
        while self.pos < len(self.src) and (self._ch().isalnum() or self._ch() == "_"):
            self._advance()
        return self.src[start:self.pos]

    def tokenize(self) -> list[Token]:
        """Tokenize the entire source, return token list ending with EOF."""
        while self.pos < len(self.src):
            self._skip_whitespace_and_comments()
            if self.pos >= len(self.src):
                break

            line, col = self.line, self.col
            ch = self._ch()

            # F-strings
            if ch == "f" and self._peek() == '"':
                self._advance()  # f
                self._advance()  # "
                val = self._read_string(is_fstring=True)
                self._emit(TT.FSTRING_LIT, val, line, col)
                continue

            # Identifiers and keywords
            if ch.isalpha() or ch == "_":
                word = self._read_ident()
                tt = KEYWORDS.get(word, TT.IDENT)
                self._emit(tt, word, line, col)
                continue

            # Numbers
            if ch.isdigit():
                tt, val = self._read_number()
                self._emit(tt, val, line, col)
                continue

            # Strings
            if ch == '"':
                self._advance()
                val = self._read_string()
                self._emit(TT.STRING_LIT, val, line, col)
                continue

            # Two-char operators
            nch = self._peek()
            two = ch + nch
            if two == ":=":
                self._advance(); self._advance()
                self._emit(TT.COLON_EQ, ":=", line, col); continue
            if two == "==":
                self._advance(); self._advance()
                self._emit(TT.EQ_EQ, "==", line, col); continue
            if two == "!=":
                self._advance(); self._advance()
                self._emit(TT.BANG_EQ, "!=", line, col); continue
            if two == "<=":
                self._advance(); self._advance()
                self._emit(TT.LT_EQ, "<=", line, col); continue
            if two == ">=":
                self._advance(); self._advance()
                self._emit(TT.GT_EQ, ">=", line, col); continue
            if two == "->":
                self._advance(); self._advance()
                self._emit(TT.ARROW, "->", line, col); continue
            if two == "=>":
                self._advance(); self._advance()
                self._emit(TT.FAT_ARROW, "=>", line, col); continue
            if two == "&&":
                self._advance(); self._advance()
                self._emit(TT.AMP_AMP, "&&", line, col); continue
            if two == "||":
                self._advance(); self._advance()
                self._emit(TT.PIPE_PIPE, "||", line, col); continue
            if two == "<<":
                self._advance(); self._advance()
                self._emit(TT.SHL, "<<", line, col); continue
            if two == ">>":
                self._advance(); self._advance()
                self._emit(TT.SHR, ">>", line, col); continue
            if two == "**":
                self._advance(); self._advance()
                self._emit(TT.STAR_STAR, "**", line, col); continue
            if two == "..":
                self._advance(); self._advance()
                self._emit(TT.DOT_DOT, "..", line, col); continue

            # Single-char
            self._advance()
            SINGLE = {
                "+": TT.PLUS, "-": TT.MINUS, "*": TT.STAR, "/": TT.SLASH,
                "%": TT.PERCENT, "<": TT.LT, ">": TT.GT,
                "&": TT.AMP, "|": TT.PIPE, "^": TT.CARET,
                "!": TT.BANG, "?": TT.QUESTION,
                "(": TT.LPAREN, ")": TT.RPAREN,
                "{": TT.LBRACE, "}": TT.RBRACE,
                "[": TT.LBRACKET, "]": TT.RBRACKET,
                ",": TT.COMMA, ";": TT.SEMICOLON,
                ":": TT.COLON, ".": TT.DOT,
                "=": TT.EQ, "_": TT.UNDERSCORE,
            }
            if ch in SINGLE:
                self._emit(SINGLE[ch], ch, line, col)
            else:
                raise LexError(f"unexpected character {ch!r}", line, col)

        self._emit(TT.EOF, "", self.line, self.col)
        return self.tokens


def lex(source: str) -> list[Token]:
    """Tokenize Mog source code."""
    return Lexer(source).tokenize()
